### PR TITLE
Add database_type to Field

### PR DESCRIPTION
**Pull request by camsaul**
Start recording the actual type of a column in the database, e.g. `VARCHAR` or `CITEXT` or `MYSTERY_FLAVOR`. This will be used to power things like Postgres enum support without having to rely on hackily dynamically modifying our type hierarchy.

### CODE REVIEW

File: src/metabase/driver/generic_sql.clj [hunk 1]
```diff
@@ -325,20 +325,29 @@
...
+  [driver column-name database-type]
+  (when-let [special-type (column->special-type driver column-name (keyword database-type))]
+    (assert (isa? special-type :type/*)
+      (str "Invalid type: " special-type))
+    special-type))
+
+(defn- describe-table-fields [^DatabaseMetaData metadata, driver, {schema :schema, table-name :name}]
+  (set (for [{database-type :type_name, column-name :column_name} (jdbc/result-set-seq (.getColumns metadata nil schema table-name nil))]
+         (merge {:name          column-name
+                 :database-type (name database-type)
```
Comment (senior) [inline src/metabase/driver/generic_sql.clj]: This will always be a string out of the database, right?

File: src/metabase/driver/mongo.clj [hunk 2]
```diff
@@ -104,22 +106,28 @@
                                  (find-nested-fields field-value nested-fields)
                                  nested-fields)))))
 
+(s/defn ^:private ^:always-validate most-common-object-type :- Class
```
Comment (senior) [inline src/metabase/driver/mongo.clj]: Probably muscle memory here on the `^:always-validate` :smiley:

**Review by senior — APPROVED** [review]
LGTM 👍

File: src/metabase/driver/mongo.clj (hunk 2 above)
Comment (camsaul) [reply src/metabase/driver/mongo.clj]: actually that's just code from before yesterday
I'll fix that

File: src/metabase/driver/generic_sql.clj (hunk 1 above)
Comment (camsaul) [reply src/metabase/driver/generic_sql.clj]: yeah. It will. good point

**Pull request merged**